url901cc5721b_shidasan_KonohaCoffee_tgz-pJ8-org_KonohaScript_KonohaJ8/src/org/KonohaScript/CodeGen/SimpleVMCodeGen.java

```diff
@@ -0,0 +1,423 @@
+
+	}
+
+	@Override
+	public void EnterLocal(LocalNode Node) {
+		AddLocalVarIfNotDefined(Node.SourceToken.ParsedText);
+	}
+
+	@Override
+	public boolean ExitLocal(LocalNode Node) {
+		push(Node.SourceToken.ParsedText);
+		return true;
+
+	}
+
+	@Override
+	public void EnterField(FieldNode Node) {
+		AddLocalVarIfNotDefined(Node.TermToken.ParsedText);
+	}
+
+	@Override
+	public boolean ExitField(FieldNode Node) {
+		String Expr = Node.TermToken.ParsedText;
+		// push(Expr + "." + Node.TypeInfo.FieldNames.get(Node.Xindex));
+		return true;
+
+	}
+
+	@Override
+	public void EnterBox(BoxNode Node) {
+		/* do nothing */
+	}
+
+	@Override
+	public boolean ExitBox(BoxNode Node) {
+		/* do nothing */return true;
+
+	}
+
+	@Override
+	public void EnterMethodCall(MethodCallNode Node) {
+		/* do nothing */
+	}
+
+	@Override
+	public boolean ExitMethodCall(MethodCallNode Node) {
+		String Params = "";
+		String methodName = "mtd";
+		int ParamSize = Node.Params.size();
+		for (int i = 0; i < ParamSize; i = i + 1) {
+			String Expr = pop();
+			if (i != 0) {
+				Params = "," + Params;
+			}
+			Params = Expr + Params;
+		}
+		String thisNode = pop();
+		push(thisNode + "." + methodName + "(" + Params + ")");
+		return true;
+
+	}
+
+	@Override
+	public void EnterAnd(AndNode Node) {
+		/* do nothing */
+	}
+
+	@Override
+	public boolean ExitAnd(AndNode Node) {
+		String Right = pop();
+		String Left = pop();
+		push(Left + " && " + Right);
+		return true;
+
+	}
+
+	@Override
+	public void EnterOr(OrNode Node) {
+		/* do nothing */
+	}
+
+	@Override
+	public boolean ExitOr(OrNode Node) {
+		String Right = pop();
+		String Left = pop();
+		push(Left + " || " + Right);
+		return true;
+
+	}
+
+	@Override
+	public void EnterAssign(AssignNode Node) {
+		AddLocalVarIfNotDefined(Node.TermToken.ParsedText);
+	}
+
+	@Override
+	public boolean ExitAssign(AssignNode Node) {
+		String Right = pop();
+		push(Node.TermToken.ParsedText + " = " + Right);
+		return true;
+
+	}
+
+	@Override
+	public void EnterLet(LetNode Node) {
+		AddLocalVarIfNotDefined(Node.TermToken.ParsedText);
+	}
+
+	@Override
+	public boolean ExitLet(LetNode Node) {
+		String Block = pop();
+		String Right = pop();
+		push(Node.TermToken.ParsedText + " = " + Right + Block);
+		return true;
+
+	}
+
+	@Override
+	public void EnterBlock(BlockNode Node) {
+		this.CurrentProgramSize = this.Program.size();
+	}
+
+	@Override
+	public boolean ExitBlock(BlockNode Node) {
+		String Exprs = "";
+		int Size = this.Program.size() - this.CurrentProgramSize;
+		for (int i = 0; i < Size; i = i + 1) {
+			String Expr = pop();
+			Exprs = Expr + ";" + Exprs;
+		}
+		push("{" + Exprs + "}");
+		return true;
+
+	}
+
+	@Override
+	public void EnterIf(IfNode Node) {
+		/* do nothing */
+	}
+
+	@Override
+	public boolean ExitIf(IfNode Node) {
+		String ElseBlock = pop();
+		String ThenBlock = pop();
+		String CondExpr = pop();
+		push("if (" + CondExpr + ") " + ThenBlock + " else " + ElseBlock);
+		return true;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 274:278, 308:312.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.